File: SiriusQuality/SiriusQuality-Priestly-Taylor/src/apsim/SiriusQuality_Priestly_Taylor/PriestlyTaylor.cs
using System;
using System.Collections.Generic;
using System.Linq;    
using Models.Core;   
namespace Models.Crop2ML;

/// <summary>
///- Name: PriestlyTaylor -Version: 1.0, -Time step: 1
///- Description:
///            * Title: evapoTranspirationPriestlyTaylor  Model
///            * Authors: Peter D. Jamieson, Glen S. Francis, Derick R. Wilson, Robert J. Martin
///            * Reference: https://doi.org/10.1016/0168-1923(94)02214-5
///            * Institution: New Zealand Institute for Crop and Food Research Ltd.,
///New Zealand Institute for Crop and Food Research Ltd.,
///New Zealand Institute for Crop and Food Research Ltd.,
///New Zealand Institute for Crop and Food Research Ltd.
///
///            * ExtendedDescription: Calculate Energy Balance
///            * ShortDescription: It uses Priestly-Taylor method
///- inputs:
///            * name: lambdaV
///                          ** description : latent heat of vaporization of water
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 10.0
///                          ** min : 0.0
///                          ** default : 2.454
///                          ** unit : MJ kg-1
///            * name: netRadiation
///                          ** description : net radiation
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLE
///                          ** max : 5000.0
///                          ** min : 0.0
///                          ** default : 1.566
///                          ** unit : MJ m-2 d-1
///            * name: psychrometricConstant
///                          ** description : psychrometric constant
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 1.0
///                          ** min : 0.0
///                          ** default : 0.66
///                          ** unit : 
///            * name: Alpha
///                          ** description : Priestley-Taylor evapotranspiration proportionality constant
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 100.0
///                          ** min : 0.0
///                          ** default : 1.5
///                          ** unit : 
///            * name: solarRadiation
///                          ** description : solar Radiation
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLE
///                          ** max : 1000.0
///                          ** min : 0.0
///                          ** default : 3.0
///                          ** unit : MJ m-2 d-1
///            * name: hslope
///                          ** description : the slope of saturated vapor pressure temperature curve at a given temperature
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLE
///                          ** max : 1000.0
///                          ** min : 0.0
///                          ** default : 0.584
///                          ** unit : hPa degC-1
///            * name: ih
///                          ** description : hour of the day if the component is hourly, -999 if the component is daily
///                          ** inputtype : variable
///                          ** parametercategory : constant
///                          ** datatype : INT
///                          ** max : 24
///                          ** min : -999
///                          ** default : -999
///                          ** unit : 
///- outputs:
///            * name: evapoTranspirationPriestlyTaylor
///                          ** description : evapoTranspiration of Priestly Taylor
///                          ** datatype : DOUBLE
///                          ** variablecategory : rate
///                          ** max : 10000.0
///                          ** min : 0.0
///                          ** unit : g m-2 d-1
/// </summary>
public class PriestlyTaylor
{

    private double _lambdaV;
    /// <summary>
    /// Gets and sets the latent heat of vaporization of water
    /// </summary>
    [Description("latent heat of vaporization of water")] 
    [Units("MJ kg-1")] 
    //[Crop2ML(datatype="DOUBLE", min=0.0, max=10.0, default=2.454, parametercategory=constant, inputtype="parameter")] 
    public double lambdaV
    {
        get { return this._lambdaV; }
        set { this._lambdaV= value; } 
    }

    private double _psychrometricConstant;
    /// <summary>
    /// Gets and sets the psychrometric constant
    /// </summary>
    [Description("psychrometric constant")] 
    [Units("")] 
    //[Crop2ML(datatype="DOUBLE", min=0.0, max=1.0, default=0.66, parametercategory=constant, inputtype="parameter")] 
    public double psychrometricConstant
    {
        get { return this._psychrometricConstant; }
        set { this._psychrometricConstant= value; } 
    }

    private double _Alpha;
    /// <summary>
    /// Gets and sets the Priestley-Taylor evapotranspiration proportionality constant
    /// </summary>
    [Description("Priestley-Taylor evapotranspiration proportionality constant")] 
    [Units("")] 
    //[Crop2ML(datatype="DOUBLE", min=0.0, max=100.0, default=1.5, parametercategory=constant, inputtype="parameter")] 
    public double Alpha
    {
        get { return this._Alpha; }
        set { this._Alpha= value; } 
    }

    private int _ih;
    /// <summary>
    /// Gets and sets the hour of the day if the component is hourly, -999 if the component is daily
    /// </summary>
    [Description("hour of the day if the component is hourly, -999 if the component is daily")] 
    [Units("")] 
    //[Crop2ML(datatype="INT", min=-999, max=24, default=-999, parametercategory=constant, inputtype="parameter")] 
    public int ih
    {
        get { return this._ih; }
        set { this._ih= value; } 
    }

    
    /// <summary>
    /// Constructor of the PriestlyTaylor component")
    /// </summary>  
    public PriestlyTaylor() { }
    
    /// <summary>
    /// Algorithm of the PriestlyTaylor component
    /// </summary>
    public void  CalculateModel(PotentialET_PTState s, PotentialET_PTState s1, PotentialET_PTRate r, PotentialET_PTAuxiliary a, PotentialET_PTExogenous ex)
    {
        double netRadiation = a.netRadiation;
        double solarRadiation = a.solarRadiation;
        double hslope = a.hslope;
        double evapoTranspirationPriestlyTaylor;
        double a_G_Rn;
        a_G_Rn = 1.00;
        if (ih != -999)
        {
            if (solarRadiation < 0.001)
            {
                a_G_Rn = 0.50;
            }
            else
            {
                a_G_Rn = 0.90;
            }
        }
        evapoTranspirationPriestlyTaylor = Math.Max(Alpha * hslope * (netRadiation / lambdaV * 1000.00) * a_G_Rn / (hslope + psychrometricConstant), 0.00);
        r.evapoTranspirationPriestlyTaylor = evapoTranspirationPriestlyTaylor;
    }
}

File: SiriusQuality/SiriusQuality-Priestly-Taylor/src/simplace/SiriusQuality_Priestly_Taylor/PriestlyTaylor.java
package net.simplace.sim.components.SiriusQuality-Priestly-Taylor;
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class PriestlyTaylor extends FWSimComponent
{
    private FWSimVariable<Double> lambdaV;
    private FWSimVariable<Double> netRadiation;
    private FWSimVariable<Double> psychrometricConstant;
    private FWSimVariable<Double> Alpha;
    private FWSimVariable<Double> solarRadiation;
    private FWSimVariable<Double> hslope;
    private FWSimVariable<Integer> ih;
    private FWSimVariable<Double> evapoTranspirationPriestlyTaylor;

    public PriestlyTaylor(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public PriestlyTaylor(){
        super();
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("lambdaV", "latent heat of vaporization of water", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"MJ kg-1", 0.0, 10.0, 2.454, this));
        addVariable(FWSimVariable.createSimVariable("netRadiation", "net radiation", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"MJ m-2 d-1", 0.0, 5000.0, 1.566, this));
        addVariable(FWSimVariable.createSimVariable("psychrometricConstant", "psychrometric constant", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0.0, 1.0, 0.66, this));
        addVariable(FWSimVariable.createSimVariable("Alpha", "Priestley-Taylor evapotranspiration proportionality constant", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"", 0.0, 100.0, 1.5, this));
        addVariable(FWSimVariable.createSimVariable("solarRadiation", "solar Radiation", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"MJ m-2 d-1", 0.0, 1000.0, 3.0, this));
        addVariable(FWSimVariable.createSimVariable("hslope", "the slope of saturated vapor pressure temperature curve at a given temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"hPa degC-1", 0.0, 1000.0, 0.584, this));
        addVariable(FWSimVariable.createSimVariable("ih", "hour of the day if the component is hourly, -999 if the component is daily", DATA_TYPE.INT, CONTENT_TYPE.constant,"", -999, 24, -999, this));
        addVariable(FWSimVariable.createSimVariable("evapoTranspirationPriestlyTaylor", "evapoTranspiration of Priestly Taylor", DATA_TYPE.DOUBLE, CONTENT_TYPE.rate,"g m-2 d-1", 0.0, 10000.0, null, this));

        return iFieldMap;
    }
    @Override
    protected void process()
    {
        double t_lambdaV = lambdaV.getValue();
        double t_netRadiation = netRadiation.getValue();
        double t_psychrometricConstant = psychrometricConstant.getValue();
        double t_Alpha = Alpha.getValue();
        double t_solarRadiation = solarRadiation.getValue();
        double t_hslope = hslope.getValue();
        Integer t_ih = ih.getValue();
        double t_evapoTranspirationPriestlyTaylor = evapoTranspirationPriestlyTaylor.getDefault();
        double a_G_Rn;
        a_G_Rn = 1.00d;
        if (t_ih != -999)
        {
            if (t_solarRadiation < 0.001d)
            {
                a_G_Rn = 0.50d;
            }
            else
            {
                a_G_Rn = 0.90d;
            }
        }
        t_evapoTranspirationPriestlyTaylor = Math.max(t_Alpha * t_hslope * (t_netRadiation / t_lambdaV * 1000.00d) * a_G_Rn / (t_hslope + t_psychrometricConstant), 0.00d);
        evapoTranspirationPriestlyTaylor.setValue(t_evapoTranspirationPriestlyTaylor, this);
    }

    @Override
    protected void init()
    {
    }
    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new PriestlyTaylor(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }
}